File: Refresh/CaseyRefresh-master/TableRefresh/Ali/AliRefreshHeaderView.m
//
//  CyRefreshHeaderDefaultView.m
//  TableRefresh
//
//  Created by Casey on 05/12/2018.
//  Copyright © 2018 Casey. All rights reserved.
//

#import "AliRefreshHeaderView.h"
#import "UIView+CyRefreshFrame.h"
#import "CyStringWidthModel.h"
#import "AliRefreshAnimationView.h"

@interface AliRefreshHeaderView (){
    
    
    BOOL isEndRefreshing;
}

@property (nonatomic, assign)CGFloat idleTextWidth;
@property (nonatomic, assign)CGFloat loadOperationTextWidth;
@property (nonatomic, assign)CGFloat willLoadTextWidth;
@property (nonatomic, assign)CGFloat loadingTextWidth;


@end

@implementation AliRefreshHeaderView{
    
    AliRefreshAnimationView *_animationView;
}


- (instancetype)initWithFrame:(CGRect)frame {
    
    self = [super initWithFrame:frame];
    if (self) {
        
        
        _animationView  = [[AliRefreshAnimationView alloc] initWithFrame:CGRectMake(0, 0, 100, self.frame.size.height)];
        [self addSubview:_animationView];
        
        self.automaticallyChangeAlpha = NO;
    }
    
    return self;
    
}


//
- (void)layoutSubviews{
    
    [super layoutSubviews];
    
    _animationView.frame = CGRectMake((self.frame.size.width-100)/2, 0, 100, self.frame.size.height);
    
    switch (self.state) {
            
        case CyRefreshStateRefreshing:
            [self loadingStatusView];
            break;
       
        default:
            break;
    }
}



- (void)beginRefreshing {
    
    [super beginRefreshing];
    UIScrollView *superView = (UIScrollView*)self.superview;
    if ([superView isKindOfClass:[UIScrollView class]]) {
        [_animationView animationScrollViewOffsetY:-self.cy_height];
    }
}


- (void)endRefreshing {
    
    isEndRefreshing = YES;
    [CATransaction flush];
    [_animationView animationEndLoading:^{
        [super endRefreshing];
        self->isEndRefreshing = NO;
    }];
    
    [CATransaction commit];
}

- (void)scrollViewMoving {
    
    UIScrollView *superView = (UIScrollView*)self.superview;
    if ([superView isKindOfClass:[UIScrollView class]]) {
        if (superView.dragging){
            
            [_animationView animationScrollViewOffsetY:superView.contentOffset.y];
            
        }else {
            
            if (self.state == CyRefreshStateRefreshing && !isEndRefreshing) {
                [_animationView animationFlush];
            }else {
                
            }
            
        }
        
    }
    
}


//正在加载
- (void)loadingStatusView {
    
    [_animationView animationFlush];
    
}




@end

File: Refresh/Refresh/Ali/AliRefreshHeaderView.swift
//
//  AliRefreshHeaderView.swift
//  Refresh
//
//  Created by Casey on 12/04/2019.
//  Copyright © 2019 nb. All rights reserved.
//

import UIKit

class AliRefreshHeaderView: RefreshHeaderBaseView {

    
    let _animationView = AliRefreshAnimationView.init(frame: CGRect.init(x: 0, y: 0, width: 100, height: RefreshHeaderHeight))
    override init(frame: CGRect) {
        
        super.init(frame: frame)
        
        self.addSubview(_animationView)
        
        self.automaticallyChangeAlpha = false
        
        
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    
    override func layoutSubviews() {
        
        super.layoutSubviews()
        
        _animationView.centerX = self.width/2
        
        // 不需要捕获状态 更具偏移位置来操作
        
    }
    
    // 开始刷新
    override func beginRefreshing() {
        
        super.beginRefreshing()
        
        if let _ = self.superview as? UIScrollView {
            
            _animationView.animationScrollViewOffsetY(-self.height)
        }
        
        
        
    }
    
    // 结束刷新
    var _isEndRefreshing = false // 正在执行结束刷新操作
    override func endRefreshing() {
        
        _isEndRefreshing = true
        
        CATransaction.flush()
        
        weak var weakSelf = self
        _animationView.animationEndLoading {
            
            weakSelf?.endrefreshFinsiAnimation()
        }
        
        CATransaction.commit()
        
    }
    
    func endrefreshFinsiAnimation()  {
        
        super.endRefreshing()
        _isEndRefreshing = false
        
    }
    
    // 滑动
    override func scrollViewMoving() {
        
        super.scrollViewMoving()
        
        
        if let scrollView = self.superview as? UIScrollView {
            
            if scrollView.isDragging {
                
                _animationView .animationScrollViewOffsetY(scrollView.contentOffset.y)
                
            }else {
                
                if self.state == .refreshing && !_isEndRefreshing { // 开始执行加载动画
                    
                    // _isEndRefreshing的条件作用：当在执行结束刷新动画的过程中，不能执行刷新动画，因为刷新结束操作没有完成
                    _animationView.animationStartLoading()
                }
                
            }
            
        }
        
        
    }
    
}
